adapters/repos/db/shard_resource_use.go

```diff
@@ -0,0 +1,151 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2022 SeMI Technologies B.V. All rights reserved.
+//
+//  CONTACT: hello@semi.technology
+//
+
+package db
+
+import (
+	"time"
+
+	"github.com/semi-technologies/weaviate/entities/storagestate"
+	"github.com/semi-technologies/weaviate/usecases/memwatch"
+)
+
+type resourceScanState struct {
+	disk *scanState
+	mem  *scanState
+}
+
+type scanState struct {
+	backoffLevel int
+	backoffs     []time.Duration
+	lastWarning  time.Time
+}
+
+func (s *scanState) getWarningInterval() time.Duration {
+	if s.backoffLevel >= len(s.backoffs) {
+		return time.Hour * 24
+	}
+
+	interval := s.backoffs[s.backoffLevel]
+
+	return interval
+}
+
+func (s *scanState) increaseWarningInterval() {
+	if s.backoffLevel < len(s.backoffs) {
+		s.backoffLevel += 1
+	}
+}
+
+func newResourceScanState() *resourceScanState {
+	backoffs := []time.Duration{
+		time.Duration(0),
+		30 * time.Second,
+		2 * time.Minute,
+		10 * time.Minute,
+		1 * time.Hour,
+		12 * time.Hour,
+	}
+
+	return &resourceScanState{
+		disk: &scanState{backoffs: backoffs},
+		mem:  &scanState{backoffs: backoffs},
+	}
+}
+
+// logs a warning if user-set threshold is surpassed
+func (s *Shard) resourceUseWarn(mon *memwatch.Monitor, du diskUse, diskPath string) {
```

Comment:
It may be more readable if there is a helper function for each resource (disk, memory, ...).

Reply:
fixed